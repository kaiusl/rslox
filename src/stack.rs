use core::slice;
use std::marker::PhantomData;
use std::{mem, ops};

use crate::value::Value;

pub const STACK_MAX: usize = 1024;

pub fn create_stack() -> [Value; STACK_MAX] {
    [Value::NIL; STACK_MAX]
}

#[derive(Debug)]
pub struct Stack<'a> {
    start: *mut Value,
    end: *mut Value,
    sp: *mut Value,
    marker: PhantomData<&'a mut [Value; STACK_MAX]>,
}

impl<'stack> Stack<'stack> {
    pub fn new(array: &'stack mut [Value; STACK_MAX]) -> Self {
        let array = array.as_mut_ptr_range();
        Self {
            start: array.start,
            end: array.end,
            sp: array.start,
            marker: PhantomData,
        }
    }

    #[inline]
    pub fn is_full(&self) -> bool {
        self.sp == self.end
    }

    #[inline]
    pub fn push(&mut self, value: Value) -> Result<(), ()> {
        if self.sp == self.end {
            return Err(());
        }

        unsafe {
            self.push_unchecked(value);
        }
        Ok(())
    }

    #[inline]
    pub unsafe fn push_unchecked(&mut self, value: Value) {
        unsafe {
            self.sp.write(value);
            self.sp = self.sp.add(1);
        }
    }

    #[inline]
    pub fn len(&self) -> usize {
        unsafe { self.sp.offset_from(self.start) as usize }
    }

    #[must_use]
    #[inline]
    pub fn is_empty(&self) -> bool {
        self.sp == self.start
    }

    #[inline]
    pub fn get(&self, index: usize) -> Option<&Value> {
        let ptr = unsafe { self.start.add(index) };
        if ptr >= self.sp {
            return None;
        }

        Some(unsafe { &*ptr })
    }

    #[inline]
    pub fn get_slice(&self, range: ops::RangeFrom<usize>) -> Option<&[Value]> {
        let start = unsafe { self.start.add(range.start) };
        if start >= self.sp {
            return None;
        }
        let len = unsafe { self.sp.offset_from(start) };
        assert!(len > 0);

        Some(unsafe { std::slice::from_raw_parts(start, len as usize) })
    }

    #[inline]
    pub fn get_mut(&mut self, index: usize) -> Option<&mut Value> {
        let ptr = unsafe { self.start.add(index) };
        if ptr >= self.sp {
            return None;
        }

        Some(unsafe { &mut *ptr })
    }

    #[inline]
    pub fn last(&self) -> Option<&Value> {
        if self.sp == self.start {
            return None;
        }

        Some(unsafe { &*self.sp.sub(1) })
    }

    pub fn swap_remove(&mut self, index: usize) -> Option<Value> {
        let ptr = unsafe { self.start.add(index) };
        if ptr >= self.sp {
            return None;
        }

        self.sp = unsafe { self.sp.sub(1) };

        unsafe { std::ptr::swap(ptr, self.sp) };
        Some(unsafe { std::ptr::read(self.sp) })
    }

    #[inline]
    pub fn pop(&mut self) -> Option<Value> {
        if self.sp == self.start {
            return None;
        }
        self.sp = unsafe { self.sp.sub(1) };
        let result = unsafe { std::ptr::read(self.sp) };
        //unsafe { std::ptr::write(self.sp, Value::Nil) }
        Some(result)
    }

    #[inline]
    pub fn pop2(&mut self) -> Option<(Value, Value)> {
        if self.len() < 2 {
            return None;
        }

        self.sp = unsafe { self.sp.sub(2) };

        Some(unsafe { (std::ptr::read(self.sp), std::ptr::read(self.sp.add(1))) })
    }

    #[inline]
    pub fn truncate(&mut self, len: usize) {
        assert!(len <= self.len());
        self.sp = unsafe { self.start.add(len) };
    }

    #[inline]
    pub fn pop_n(&mut self, count: usize) {
        assert!(count <= self.len());

        if count == 0 {
            return;
        }

        self.sp = unsafe { self.sp.sub(count) };
    }

    pub fn iter(&self) -> slice::Iter<'_, Value> {
        unsafe { slice::from_raw_parts(self.start, self.len()).iter() }
    }
}

impl ops::Index<usize> for Stack<'_> {
    type Output = Value;
    #[inline]
    fn index(&self, index: usize) -> &Self::Output {
        self.get(index).unwrap()
    }
}

impl ops::Index<ops::RangeFrom<usize>> for Stack<'_> {
    type Output = [Value];
    #[inline]
    fn index(&self, index: ops::RangeFrom<usize>) -> &Self::Output {
        self.get_slice(index).unwrap()
    }
}

impl ops::IndexMut<usize> for Stack<'_> {
    #[inline]
    fn index_mut(&mut self, index: usize) -> &mut Self::Output {
        self.get_mut(index).unwrap()
    }
}
