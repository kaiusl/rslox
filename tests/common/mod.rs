#![allow(
    dead_code,
    unused_macros,
    unused_imports,
    reason = "these methods are used in other modules of integration tests, but compiler doesn't see it"
)]

pub(crate) fn test_file(prefix: &str, root: impl AsRef<str>, path: impl AsRef<std::path::Path>) {
    let root = root.as_ref();
    miette::set_hook(Box::new(|_| {
        Box::new(
            miette::MietteHandlerOpts::new()
                .color(false)
                .wrap_lines(false)
                .break_words(false)
                .build(),
        )
    }))
    .ok();

    let path = path.as_ref();
    let src = std::fs::read_to_string(path).unwrap();
    let mut out = Vec::<u8>::new();
    let mut outerr = Vec::<u8>::new();
    let mut stack = create_stack();
    let mut vm = rslox::Vm::with_output(&mut stack, &mut out, &mut outerr);
    if vm.compile(&src).is_ok() {
        vm.run(&src);
    };

    let root_dir = &root[2..];
    let mut settings = insta::Settings::clone_current();
    let mut suffix = String::from(prefix);
    for c in path.components() {
        let str = c.as_os_str().to_str().unwrap();
        if !str.is_empty() && str != "." && str != root_dir {
            if !suffix.is_empty() {
                suffix.push('@');
            }
            suffix.push_str(c.as_os_str().to_str().unwrap());
        }
    }

    #[cfg(not(miri))]
    {
        settings.set_snapshot_suffix(suffix);
        settings.set_input_file(path);
        settings.set_description(src);
        settings.set_prepend_module_to_snapshot(false);
        settings.set_omit_expression(true);
        settings.set_snapshot_path("../snapshots");
        let _quard = settings.bind_to_scope();
        insta::assert_snapshot!(format!(
            "OUT: \n{}\n===\nERR: \n{}",
            String::from_utf8(out).unwrap(),
            String::from_utf8(outerr).unwrap()
        ))
    }
}

// Note that we append _ to the test name so that we can use Rust keywords as $file
macro_rules! test_dir {
    ($prefix:ident; $type:tt; $root:literal; $($file:tt),+ $(,)?) => {
        ::paste::paste!{
            #[::rstest::rstest]
            $(
                #[case::[<$file _>](concat!(stringify!($file), ".lox"))]
            )+
            fn [<test_ $prefix _ $type>](#[case] path: &str) {
                test_file(::std::stringify!($prefix), $root, ::std::format!("{}/{}/{}", $root, ::std::stringify!($type), path));
            }
        }
    };
    ($prefix:ident; $root:literal; $($file:tt),+ $(,)?) => {
        ::paste::paste!{
            #[::rstest::rstest]
            $(
                #[case::[<$file _>](concat!(stringify!($file), ".lox"))]
            )+
            fn [< test_ $prefix  _root>](#[case] path: &str) {
                test_file(::std::stringify!($prefix), $root, ::std::format!("{}/{}", $root,  path));
            }
        }
    };
}

use rslox::stack::create_stack;
pub(crate) use test_dir;
