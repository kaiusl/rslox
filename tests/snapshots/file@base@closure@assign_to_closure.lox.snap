---
source: tests/common/mod.rs
description: "var f;\nvar g;\n\n{\n  var local = \"local\";\n  fun f_() {\n    print local;\n    local = \"after f\";\n    print local;\n  }\n  f = f_;\n\n  fun g_() {\n    print local;\n    local = \"after g\";\n    print local;\n  }\n  g = g_;\n}\n\nf();\n// expect: local\n// expect: after f\n\ng();\n// expect: after f\n// expect: after g\n"
input_file: crafting_interpreters_test_files/closure/assign_to_closure.lox
---
OUT: 
local
after f
after f
after g

===
ERR:
